src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -133,16 +133,40 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
                     collection: out collection);
             }
 
+            // TODO: in both nongeneric and generic forks below, we should unify both "custom tasklike" and "Task/Task<T>"
+            // down the same custom-tasklike path. (Should mscorlib ever device to put [AsyncBuilder] on Task then that
+            // will happen anyway).
+
             if (method.IsTaskReturningAsync(F.Compilation))
             {
                 var returnType = (NamedTypeSymbol)method.ReturnType;
                 NamedTypeSymbol builderType;
-                MethodSymbol createBuilderMethod;
-                PropertySymbol taskProperty;
-                bool customBuilder = returnType.IsCustomTaskType(out builderType, out createBuilderMethod);
+                MethodSymbol createBuilderMethod = null;
+                PropertySymbol taskProperty = null;
+
+                object builderArgument;
+                bool customBuilder = returnType.IsCustomTaskType(out builderArgument);
                 if (customBuilder)
                 {
-                    taskProperty = GetCustomTaskProperty(F, builderType);
+                    builderType = builderArgument as NamedTypeSymbol;
+                    if (builderType?.IsErrorType() == true)
+                    {
+                    }
+                    else if ((object)builderType == null || 
+                             builderType.SpecialType == SpecialType.System_Void ||
+                             builderType.DeclaredAccessibility != returnType.DeclaredAccessibility ||
+                             builderType.IsGenericType)
+                    {
+                        // might be null if type isn't a named type, e.g. [AsyncBuilder(typeof(object[]))]
+                        var diagnostic = new CSDiagnostic(
+                            new CSDiagnosticInfo(ErrorCode.ERR_BadAsyncReturn), F.Syntax.Location);
+                        F.Diagnostics.Add(diagnostic);
```

Comment:
Fixed in https://github.com/dotnet/roslyn/pull/13405/commits/eef0c70b2966698aa71be1063a920290e5778f51#diff-995d1bf33206eef8c980dbd08b8601b0R295